File: apps/android/core/playback/src/main/kotlin/app/storyarc/core/playback/SkipPreferences.kt
package app.storyarc.core.playback

import android.content.Context
import android.content.SharedPreferences

/**
 * How far the listener asked a skip to go, kept where the service can read it.
 *
 * `audio-playback` asks for an interval "the listener can configure", which is the clause
 * that makes this a store rather than two constants.
 *
 * **Here rather than in `:core:persistence`**, for [PlaybackMemory]'s reason exactly:
 * [PlaybackService] sets the decoder's own seek increments and labels the notification's two
 * outer buttons, and a service the system has just started to answer the shade carousel has
 * no scope, no database and no time — while `:core:playback` deliberately does not depend on
 * the library's store at all. A second small preferences file is the cost of that boundary,
 * and it is the same cost the resumption memory already pays.
 *
 * Read synchronously, because both readers are already drawing something when they ask.
 */
class SkipPreferences internal constructor(private val preferences: SharedPreferences) {

    /** What the listener chose, or the defaults on a device that has never been asked. */
    fun intervals(): SkipIntervals {
        val back = preferences.getInt(BACK_SECONDS, 0)
        val forward = preferences.getInt(FORWARD_SECONDS, 0)
        // Zero is the answer for a key that was never written, and it is also the one value
        // a skip may never have — a control that moves nothing. So an unset pair is the
        // default pair rather than a stopped one. `SharedPreferences` offers no sentinel
        // beyond the default it is handed, which is why this is read as a pair.
        if (back <= 0 || forward <= 0) return SkipIntervals.DEFAULT
        return SkipIntervals.of(back, forward)
    }

    /** Remembers a choice. Both directions together: they are one setting to a listener. */
    fun remember(intervals: SkipIntervals) {
        preferences.edit()
            .putInt(BACK_SECONDS, intervals.backSeconds)
            .putInt(FORWARD_SECONDS, intervals.forwardSeconds)
            .apply()
    }

    /** Back to the defaults. */
    fun forget() {
        preferences.edit().clear().apply()
    }

    companion object {
        internal const val FILE = "app.storyarc.playback.skip"
        internal const val BACK_SECONDS = "backSeconds"
        internal const val FORWARD_SECONDS = "forwardSeconds"

        fun open(context: Context): SkipPreferences = SkipPreferences(
            context.applicationContext.getSharedPreferences(FILE, Context.MODE_PRIVATE),
        )
    }
}

File: apps/ios/Packages/StoryArcKit/Sources/Persistence/SkipPreferences.swift
public import Foundation

public import Playback

/// How far one press of a skip control moves, remembered across launches.
///
/// **`audio-playback` asks for an interval "the listener can configure", and iOS had no way to
/// configure it.** `PlayerCentre.skipIntervals` existed, was read by the player, the lock screen
/// and the timeline — and had **no setter anywhere in the app**, so every listener got the
/// product defaults for ever and the requirement's second clause was unmet without anything
/// failing. This is that setter's other half.
///
/// Android's `SkipPreferences` is the contract this mirrors, case for case rather than in
/// shape: the two platforms must agree on what an unset value means, or a listener with both
/// gets two answers to one setting.
/// Not `Sendable`, and deliberately so: `UserDefaults` is not, and the neighbouring
/// `PlaybackPreferences` makes the same choice for the same reason. A store is constructed where
/// it is used rather than passed across an isolation boundary.
public struct SkipPreferences {

    private let defaults: UserDefaults
    private let prefix = "app.storyarc.playback.skip"

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// What the listener chose, or the defaults on a device that has never been asked.
    ///
    /// **Read as a pair, and a half-written pair is the defaults.** `UserDefaults` answers `0`
    /// for a key that was never written, and zero is also the one value a skip may never have —
    /// a control that moves nothing is a control that does nothing. There is no sentinel that
    /// separates the two, so an unusable half makes the whole reading unusable and the defaults
    /// stand. Android's store says the same thing about `SharedPreferences`, for the same
    /// reason.
    public func intervals() -> SkipIntervals {
        let back = defaults.double(forKey: key("back"))
        let forward = defaults.double(forKey: key("forward"))
        guard back > 0, forward > 0 else { return .default }
        return SkipIntervals(back: back, forward: forward)
    }

    /// Remembers a choice. Both directions together: to a listener they are one setting.
    public func remember(_ intervals: SkipIntervals) {
        defaults.set(intervals.back, forKey: key("back"))
        defaults.set(intervals.forward, forKey: key("forward"))
    }

    /// Back to the defaults.
    public func forget() {
        defaults.removeObject(forKey: key("back"))
        defaults.removeObject(forKey: key("forward"))
    }

    private func key(_ direction: String) -> String { "\(prefix).\(direction)" }
}
